### PR TITLE
[Log] dump stacktrace from glog lib

### PR DESCRIPTION
**Pull request by ashione**
Export a new dump stacktrace interface from glog, which will be helpful to spdlog or collecting call stack manually.

We make a comparison between glog and spdlog and find that only stacktrace is unsupported in spdlog, so this PR will reuse glog stacktrace and symbols for any module in ray.

Reference doc: https://docs.google.com/document/d/1pEXTLOhTFu5-va14Wj-tGoJD9tt_M_-JRMtxqd1JCyI/edit?usp=sharing 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
@edoakes Context: https://ray-distributed.slack.com/archives/C016MPZE1L1/p1602034725001200

File: thirdparty/patches/glog-dump-stacktrack.patch [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+--- src/logging.cc
++++ src/logging.cc
+@@ -2336,4 +2336,11 @@ void DisableLogCleaner() {
+   log_cleaner_enabled_ = false;
+ }
+ 
++std::string GetStackTraceToString() {
++  // Collect stacktrace from utilities.
++  std::string stacktrace;
++  DumpStackTraceToString(&stacktrace, 0);
```
Comment (rkooo567) [inline thirdparty/patches/glog-dump-stacktrack.patch:26]: How does this work? Does it mean we won't dump any stacktrace? (depth 0)

File: src/ray/util/logging.cc [hunk 2]
```diff
@@ -47,6 +47,14 @@
 
 namespace ray {
 
+std::string GetCallTrace() {
+  std::string return_message = "Cannot get callstack information.";
+#ifdef RAY_USE_GLOG
+  return google::GetStackTraceToString();
```
Comment (rkooo567) [inline src/ray/util/logging.cc:53]: Does it still require us to start glog? How will this look like when it is enabled with spdlog? Are we going to turn both on, but use glog only for collecting stacktrace?

Comment (ashione) [reply src/ray/util/logging.cc:53]: > Does it still require us to start glog? How will this look like when it is enabled with spdlog? Are we going to turn both on, but use glog only for collecting stacktrace?

We do not need start glog any more, just call its interface.

File: thirdparty/patches/glog-dump-stacktrack.patch (hunk 1 above)
Comment (ashione) [reply thirdparty/patches/glog-dump-stacktrack.patch:26]: Depth means stack skip count. For example, we call f1->f2->f3 and full stack view is {CallStack, f1, f2, f3} with 0 depth. If the given depth is 1, stack sequence will be {f1, f2, f3}. In short,  no stack frame will be skipped if we use DumpStackTraceToString(&stacktrace, 0);

Comment (ashione) [reply thirdparty/patches/glog-dump-stacktrack.patch:26]: BTW, max stack size is 32 by default.

Comment (rkooo567) [reply thirdparty/patches/glog-dump-stacktrack.patch:26]: Ah, I see interesting. Can you write a comment about it?

**Review by rkooo567 — APPROVED** [review]

File: src/ray/util/logging.h [hunk 3]
```diff
@@ -44,6 +44,8 @@ enum { ERROR = 0 };
 
 namespace ray {
 
+std::string GetCallTrace();
```
Comment (rkooo567) [inline src/ray/util/logging.h:47]: Add a docstring to explain why we need this.

File: src/ray/util/logging.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/util/logging.cc:53]: We still need to setup glog signal handler to log on various signals (e.g., SIGSEGV) right?

**Comment by rkooo567** [comment]
Please address a couple last comments.

File: src/ray/util/logging.cc (hunk 2 above)
Comment (ashione) [reply src/ray/util/logging.cc:53]: Nice catch. We may set up a signal handler, but it's not for glog. I'm not sure where signal handler will be used.

**Comment by ashione** [comment]
All travis ci are green, @rkooo567 could you check it again.

File: src/ray/util/logging.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/util/logging.cc:53]: I think we are currently using `google::InstallFailureSignalHandler();`. We can have further discussion inside the spdlog PR.

**Pull request merged**